addons/website/models/website_snippet_filter.py

```diff
@@ -99,21 +106,48 @@ def _filter_records_to_dict_values(self, records):
 
         self.ensure_one()
         values = []
-        model = self.env[self.filter_id.model_id] if self.filter_id else self.env[self.action_server_id.model_id.model]
+        model = self.env[self.filter_id.model_id]
+        Website = self.env['website']
         for record in records:
-            data = {
-                'fields': OrderedDict({}),
-                'image_fields': OrderedDict({}),
-            }
-
+            data = self._get_rendering_data_structure()
             for field_name in self.field_names.split(","):
+                field_name, _, field_widget = field_name.partition(":")
                 field = model._fields.get(field_name)
-                data_type = 'image_fields' if field.type == 'binary' else 'fields'
-                if data_type == 'image_fields' and self.filter_id:
-                    data[data_type][field_name] = self.env['website'].image_url(record, field_name)
+                field_widget = field_widget or field.type
+                if field.type == 'binary':
+                    data['image_fields'][field_name] = Website.image_url(record, field_name)
+                elif field_widget == 'image':
+                    data['image_fields'][field_name] = record[field_name]
+                elif field_widget == 'monetary':
+                    FieldMonetary = self.env['ir.qweb.field.monetary']
+                    website_currency = self._get_website_currency()
```

Comment:
can be moved in if, or even directly in display_currency, no need to compute it if you don't entry in if